### PR TITLE
JAMES-2905 ES + Nginx connection tests

### PR DESCRIPTION
**Pull request by trantienduchn**

### CODE REVIEW

File: backends-common/elasticsearch/src/test/java/org/apache/james/backends/es/DockerElasticSearch.java [hunk 1]
```diff
@@ -55,85 +76,188 @@ static ElasticSearchAPI from(Host esHttpHost) {
...
-        return Host.from(getIp(), getHttpPort());
+    class WithAuth implements DockerElasticSearch {
+
+        private static final String DEFAULT_USERNAME = "elasticsearch";
+        private static final String DEFAULT_PASSWORD = "secret";
+        private static final ElasticSearchConfiguration.Credential DEFAULT_CREDENTIAL =
+            ElasticSearchConfiguration.Credential.of(DEFAULT_USERNAME, DEFAULT_PASSWORD);
+
+        private static final String NGINX_SERVICE = "nginx";
+        private static final int COMPILATION_BY_PASS_DUMMY_PORT = 0;
```
Comment (remk) [inline backends-common/elasticsearch/src/test/java/org/apache/james/backends/es/DockerElasticSearch.java]: ???
why compilation ?
bypass or by passing ?

Comment (trantienduchn) [reply backends-common/elasticsearch/src/test/java/org/apache/james/backends/es/DockerElasticSearch.java]: IDK how to name it, just a dummy value to pass to the method. It does nothing

File: backends-common/elasticsearch/src/test/resources/auth-es/nginx-conf/passwd [hunk 2]
```diff
@@ -0,0 +1 @@
+elasticsearch:$apr1$m5FvKpw7$EZ9UEUCilHyKTIqQ57nWG1
```
Comment (chibenwa) [inline backends-common/elasticsearch/src/test/resources/auth-es/nginx-conf/passwd]: Please have a comment, or a README next to this file giving the password in plain text...

File: backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/ClientProvider.java [hunk 3]
```diff
@@ -46,14 +53,76 @@
 
 public class ClientProvider implements Provider<RestHighLevelClient> {
 
+    private static class HttpAsyncClientConfigure {
```
Comment (chibenwa) [inline backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/ClientProvider.java]: Configure is a verb, we should use nomn as class names.

File: backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/ClientProvider.java [hunk 4]
```diff
@@ -46,14 +53,76 @@
...
+        }
+
+        private void configureSSL(HttpAsyncClientBuilder builder) {
+            if (configuration.getHostScheme().equals(HostScheme.HTTP)) {
+                return;
+            }
+
+            try {
+                SSLContext sslContext = SSLContext.getInstance("SSL");
+                sslContext.init(null, TRUST_ALL, new java.security.SecureRandom());
```
Comment (chibenwa) [inline backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/ClientProvider.java]: what is null?

File: backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/ClientProvider.java [hunk 5]
```diff
@@ -19,22 +19,29 @@
 package org.apache.james.backends.es;
```
Comment (chibenwa) [inline backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/ClientProvider.java]: As a side note, loading self signed certificates, one way or another will need to be supported.

**Comment by trantienduchn** [comment]
```
[81f143e4c84d789143e955680dd1049cfd446322] .IOError: [Errno 2] No such file or directory: '/james-project/backends-common/elasticsearch/src/test/resources/auth-es/docker-compose.yml'
...
[81f143e4c84d789143e955680dd1049cfd446322] [ERROR] Errors: 
[81f143e4c84d789143e955680dd1049cfd446322] [ERROR]   ClientProviderImplConnectionAuthESTest » ContainerLaunch Container startup fai...
[81f143e4c84d789143e955680dd1049cfd446322] [ERROR]   ClientProviderImplConnectionAuthESWithSSLEnabledTest » ContainerLaunch Contain...
```

File: src/site/xdoc/server/config-elasticsearch.xml [hunk 6]
```diff
@@ -200,6 +189,49 @@
...
 
+    <section name="SSL Trusting Configuration">
+
+        <p>
+            In some cases, you want to secure ES to protect it from unauthorized requests,
+            assuming with the ES is using <strong>https</strong> with a self signed certificate.
+            Which means you should configure the ES RestHighLevelClient to trust your self signed certificate.
+
+            There are two ways on client side: ignoring SSL check or configure to trust the server's certificate.
+            In case you want to ignore the SSL check, simply, just don't specify below options. Otherwise, configuring the trust
```
Comment (chibenwa) [inline src/site/xdoc/server/config-elasticsearch.xml]: I thing we get something wrong here. Insecure by default is convenient, but is also the breach for mis-configurations.

We should:
 - Rely on JVM certificate by default
 - Have an option to effectively disable trust (that defaults to **false**
 - Have a way to load the trustStore...

File: backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/ClientProvider.java [hunk 7]
```diff
@@ -18,36 +18,110 @@
...
+
+            return builder;
+        }
+
+        private void configureSSL(HttpAsyncClientBuilder builder) {
+            if (configuration.getHostScheme().equals(HostScheme.HTTP)) {
+                return;
+            }
+
+            try {
```
Comment (chibenwa) [inline backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/ClientProvider.java]: We should fail here if the scheme is not HTTPS (avoids errors when adding some new HostScheme in the future...)

File: backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/ClientProvider.java [hunk 8]
```diff
@@ -18,36 +18,110 @@
...
+        }
+
+        private SSLContext sslContext() throws KeyStoreException, NoSuchAlgorithmException, KeyManagementException {
+            SSLContextBuilder sslContextBuilder = new SSLContextBuilder();
+
+            return configuration.getSslTrustStore()
+                .map(Throwing
+                    .<SSLTrustStore, SSLContextBuilder>function(trustStore -> sslContextBuilder
+                    .loadTrustMaterial(new File(trustStore.getFilePath()), trustStore.getPassword().toCharArray()))
+                    .sneakyThrow())
```
Comment (chibenwa) [inline backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/ClientProvider.java]: These sneakythrow are not useful as the exception is propagated literally 10 lines above. We can be more more fluent.

File: backends-common/elasticsearch/src/test/java/org/apache/james/backends/es/ElasticSearchConfigurationTest.java [hunk 9]
```diff
@@ -36,6 +40,76 @@
...
+
+    @Nested
+    class SSLTrustStoreTest {
+
+        @Test
+        void shouldMatchBeanContact() {
+            EqualsVerifier.forClass(SSLTrustStore.class)
+                .verify();
+        }
+
```
Comment (chibenwa) [inline backends-common/elasticsearch/src/test/java/org/apache/james/backends/es/ElasticSearchConfigurationTest.java]: Does it make sens to configure a trust store with an HTTP scheme?

File: backends-common/elasticsearch/src/test/java/org/apache/james/backends/es/ElasticSearchConfigurationTest.java [hunk 10]
```diff
@@ -36,6 +40,76 @@
...
+        @Test
+        void fromPropertiesShouldThrowWhenOnlyTrustStorePathProvided() throws Exception {
+            PropertiesConfiguration configuration = new PropertiesConfiguration();
+            configuration.addProperty("elasticsearch.hosts", "127.0.0.1");
+
+            configuration.addProperty("elasticsearch.hostScheme.https.trustStorePath", "/home/james/ServerTrustStore.jks");
+
+            assertThatThrownBy(() -> ElasticSearchConfiguration.fromProperties(configuration))
+                .isInstanceOf(NullPointerException.class)
+                .hasMessage("password cannot be null when filePath is specified");
```
Comment (chibenwa) [inline backends-common/elasticsearch/src/test/java/org/apache/james/backends/es/ElasticSearchConfigurationTest.java]: `password` should rather be `elasticsearch.hostScheme.https.trustStorePassword`

File: backends-common/elasticsearch/src/test/java/org/apache/james/backends/es/ElasticSearchConfigurationTest.java [hunk 11]
```diff
@@ -36,6 +40,76 @@
...
+        @Test
+        void fromPropertiesShouldThrowWhenOnlyTrustStorePasswordProvided() throws Exception {
+            PropertiesConfiguration configuration = new PropertiesConfiguration();
+            configuration.addProperty("elasticsearch.hosts", "127.0.0.1");
+
+            configuration.addProperty("elasticsearch.hostScheme.https.trustStorePassword", "secret");
+
+            assertThatThrownBy(() -> ElasticSearchConfiguration.fromProperties(configuration))
+                .isInstanceOf(NullPointerException.class)
+                .hasMessage("filePath cannot be null when password is specified");
```
Comment (chibenwa) [inline backends-common/elasticsearch/src/test/java/org/apache/james/backends/es/ElasticSearchConfigurationTest.java]: `filePath` should rather be `elasticsearch.hostScheme.https.trustStorePath`

File: src/site/xdoc/server/config-elasticsearch.xml (hunk 6 above)
Comment (trantienduchn) [reply src/site/xdoc/server/config-elasticsearch.xml]: Right, thanks!

**Comment by trantienduchn** [comment]
```
[f17643bd7ff70b99bb49ca92dc56df3d73d106b5] [ERROR] org.apache.james.backends.es.ClientProviderImplConnectionAuthESTest  Time elapsed: 78.161 s  <<< ERROR!
[f17643bd7ff70b99bb49ca92dc56df3d73d106b5] org.testcontainers.containers.ContainerLaunchException: Container startup failed
[f17643bd7ff70b99bb49ca92dc56df3d73d106b5] Caused by: org.rnorth.ducttape.RetryCountExceededException: Retry limit hit with exception
[f17643bd7ff70b99bb49ca92dc56df3d73d106b5] Caused by: org.testcontainers.containers.ContainerLaunchException: Could not create/start container
[f17643bd7ff70b99bb49ca92dc56df3d73d106b5] Caused by: org.testcontainers.containers.ContainerLaunchException: Timed out waiting for container port to open (192.168.0.1 ports: [59561] should be listening)
```

**Comment by trantienduchn** [comment]
same
```
[134f9b8aa7fb346e717bab4594a34e02dffb889f] [ERROR] org.apache.james.backends.es.ClientProviderImplConnectionAuthESWithSSLEnabledTest  Time elapsed: 77.922 s  <<< ERROR!
[134f9b8aa7fb346e717bab4594a34e02dffb889f] org.testcontainers.containers.ContainerLaunchException: Container startup failed
[134f9b8aa7fb346e717bab4594a34e02dffb889f] Caused by: org.rnorth.ducttape.RetryCountExceededException: Retry limit hit with exception
[134f9b8aa7fb346e717bab4594a34e02dffb889f] Caused by: org.testcontainers.containers.ContainerLaunchException: Could not create/start container
[134f9b8aa7fb346e717bab4594a34e02dffb889f] Caused by: org.testcontainers.containers.ContainerLaunchException: Timed out waiting for container port to open (192.168.0.1 ports: [44423] should be listening)
```

**Comment by trantienduchn** [comment]
:green_apple:

File: dockerfiles/run/guice/cassandra-ldap/destination/conf/elasticsearch.properties [hunk 12]
```diff
@@ -25,8 +25,22 @@
...
-# Optional. Only http or https are accepted, default is http
-# elasticsearch.hostScheme=http
+# Optional, default is `default`
+# Choosing the SSL check strategy when using https scheme
+# default: Use the default SSL TrustStore of the system.
+# ignore: Ignore SSL Validation check (not recommended).
+# override: Override the SSL Context to use a custome TrustStore containing ES server's certificate.
+# elasticsearch.hostScheme.https.sslValidationStrategy=default
+
+# Optional. Required when using 'https' scheme and 'override' sslValidationStrategy
```
Comment (rouazana) [inline dockerfiles/run/guice/cassandra-ldap/destination/conf/elasticsearch.properties]: why did you remove the hostScheme options then? (https://github.com/linagora/james-project/pull/2855/commits/dc05b722e2adc7f628ff406edd3bc1d1b6985bab#diff-ab4c2c6737d5ca0469265cb4b172a634L28 )

File: dockerfiles/run/guice/cassandra-rabbitmq/destination/conf/elasticsearch.properties [hunk 13]
```diff
@@ -24,8 +24,22 @@
 elasticsearch.masterHost=elasticsearch
 elasticsearch.port=9200
 
-# Optional. Only http or https are accepted, default is http
-# elasticsearch.hostScheme=http
```
Comment (rouazana) [inline dockerfiles/run/guice/cassandra-rabbitmq/destination/conf/elasticsearch.properties]: idem, you should not remove that, no?

File: src/site/xdoc/server/config-elasticsearch.xml [hunk 14]
```diff
@@ -186,6 +190,66 @@
...
+            <dd>
+                default: Use the default SSL TrustStore of the system.
+                ignore: Ignore SSL Validation check (not recommended).
+                override: Override the SSL Context to use a custome TrustStore containing ES server's certificate.
+            </dd>
+        </dl>
+
+        <p>
+            In some cases, you want to secure ES to protect it from unauthorized requests,
+            assuming with the ES is using <strong>https</strong> with a self signed certificate.
```
Comment (rouazana) [inline src/site/xdoc/server/config-elasticsearch.xml]: we need a little more explanation here: I'm not sure having nginx in front of ES is a so common practice. So can you explain a little our setup?

File: backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/ElasticSearchConfiguration.java [hunk 15]
```diff
@@ -98,6 +103,115 @@ public final int hashCode() {
...
+        public enum SSLValidationStrategy {
+            DEFAULT,
+            IGNORE,
+            OVERRIDE;
+
+            static SSLValidationStrategy from(String rawValue) {
+                Preconditions.checkNotNull(rawValue);
+
+                return Stream.of(values())
+                    .filter(strategy -> strategy.name().equalsIgnoreCase(rawValue))
```
Comment (mbaechler) [inline backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/ElasticSearchConfiguration.java]: please take care of the locale when using case-insensitive methods

File: backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/ElasticSearchConfiguration.java [hunk 16]
```diff
@@ -98,6 +103,115 @@ public final int hashCode() {
...
+        public static class SSLTrustStore {
+
+            public static SSLTrustStore of(String filePath, String password) {
+                return new SSLTrustStore(filePath, password);
+            }
+
+            private final String filePath;
+            private final String password;
+
+            private SSLTrustStore(String filePath, String password) {
```
Comment (mbaechler) [inline backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/ElasticSearchConfiguration.java]: Why don't we use a File for the file? Why don't we use char[] for the password?

File: backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/ElasticSearchConfiguration.java [hunk 17]
```diff
@@ -98,6 +103,115 @@ public final int hashCode() {
...
+            return new SSLTrustConfiguration(OVERRIDE, Optional.of(sslTrustStore));
+        }
+
+        private final SSLValidationStrategy strategy;
+        private final Optional<SSLTrustStore> trustStore;
+
+        private SSLTrustConfiguration(SSLValidationStrategy strategy, Optional<SSLTrustStore> trustStore) {
+            Preconditions.checkNotNull(strategy);
+            Preconditions.checkNotNull(trustStore);
+            Preconditions.checkArgument(strategy != OVERRIDE || trustStore.isPresent(), "OVERRIDE strategy requires trustStore to be present");
```
Comment (mbaechler) [inline backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/ElasticSearchConfiguration.java]: please try to avoid negative conditions. The code here should match the message and it's obviously not the case

File: backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/ElasticSearchConfiguration.java [hunk 18]
```diff
@@ -175,6 +291,16 @@ public Builder credential(Optional<Credential> credential) {
             return this;
         }
 
+        public Builder sslTrustConfiguration(SSLTrustConfiguration sslTrustConfiguration) {
+            this.sslTrustConfiguration = Optional.ofNullable(sslTrustConfiguration);
```
Comment (mbaechler) [inline backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/ElasticSearchConfiguration.java]: why should we support null?

File: dockerfiles/run/guice/cassandra-ldap/destination/conf/elasticsearch.properties (hunk 12 above)
Comment (trantienduchn) [reply dockerfiles/run/guice/cassandra-ldap/destination/conf/elasticsearch.properties]: sorry, my copy paste mistake

File: backends-common/elasticsearch/src/test/java/org/apache/james/backends/es/DockerElasticSearch.java [hunk 19]
```diff
@@ -19,33 +19,98 @@
...
 
     interface ElasticSearchAPI {
 
-        static ElasticSearchAPI from(Host esHttpHost) {
-            return Feign.builder()
-                .logger(new Slf4jLogger(ElasticSearchAPI.class))
-                .logLevel(Logger.Level.FULL)
-                .target(ElasticSearchAPI.class, "http://" + esHttpHost.getHostName() + ":" + esHttpHost.getPort());
+        class Builder {
+            private static final HostnameVerifier ACCEPT_ANY_HOST = (hostname1, sslSession) -> true;
```
Comment (mbaechler) [inline backends-common/elasticsearch/src/test/java/org/apache/james/backends/es/DockerElasticSearch.java]: hostname1?

File: backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/ClientProvider.java [hunk 20]
```diff
@@ -46,14 +58,103 @@
...
+                    throw new NotImplementedException(
+                        String.format("unrecognized hostScheme '%s'", scheme.name()));
+            }
+        }
+
+        private void configureSSLOptions(HttpAsyncClientBuilder builder) {
+            try {
+                builder
+                    .setSSLContext(sslContext())
+                    .setSSLHostnameVerifier(NoopHostnameVerifier.INSTANCE);
```
[the remark this answers is not in the document]
Comment (trantienduchn) [reply backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/ClientProvider.java]: one more cònfiguration item about it? if you just worry about the name, i can very well create a HostNameVerifier implemetation will a less bad than Noop...

[the remark this answers is not in the document]
Comment (rouazana) [reply backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/ClientProvider.java]: I don't worry about the name, I worry about the implementation. Is this NoopHostnameVerifier used in every case? If yes it's a very bad idea from a security point of view. If it is used only in the "ignore" case, it's pretty normal (but I did not understand that reading the code).

File: src/site/xdoc/server/config-elasticsearch.xml (hunk 14 above)
Comment (trantienduchn) [reply src/site/xdoc/server/config-elasticsearch.xml]: I prefer to put those steps into how-to documentation. do you agree?

Comment (rouazana) [reply src/site/xdoc/server/config-elasticsearch.xml]: I don't speak about the steps of starting ES with nginx base authentication, I'm just speaking of explaining that we don't use native (non-free?) ES authentication, but instead nginx (or reverse-proxy) based authentication.

Comment (trantienduchn) [reply src/site/xdoc/server/config-elasticsearch.xml]: I’m not sure but I think it can work well with a paid ES + xpack. Because I used nothing but just options provided by the RestHighLevelClient API. I mean from the client side, the configuration looks the same regardless what kind of ES. So why do you want to explain it, Im sorry I dont get it

Comment (rouazana) [reply src/site/xdoc/server/config-elasticsearch.xml]: in this case it's ok, thank you

File: backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/ClientProvider.java (hunk 20 above)
[the remark this answers is not in the document]
Comment (trantienduchn) [reply backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/ClientProvider.java]: alright, I added an easy to review implementation of the `HostnameVerifier`

File: backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/ElasticSearchConfiguration.java (hunk 17 above)
Comment (trantienduchn) [reply backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/ElasticSearchConfiguration.java]: actually it's. I just simplified the matching condition when `trategy == OVERRIDE && !trustStore.isPresent()` then throw. Had to negate it to be able to use Preconditions. I have test to ensure that. `ElasticSearchConfigurationTest.WhenOverride.fromPropertiesShouldThrowWhenTrustStoreIsNotProvided()`

**Comment by trantienduchn** [comment]
```
[2d756dc958c06342645a5c7a7796017e8091390a] [ERROR] Failed to execute goal on project james-json: Could not resolve dependencies for project org.apache.james:james-json:jar:3.5.0-SNAPSHOT: Failed to collect dependencies at com.fasterxml.jackson.datatype:jackson-datatype-guava:jar:2.9.9 -> com.google.guava:guava:jar:25.1-jre -> org.codehaus.mojo:animal-sniffer-annotations:jar:1.14: Failed to read artifact descriptor for org.codehaus.mojo:animal-sniffer-annotations:jar:1.14: Could not transfer artifact org.codehaus.mojo:animal-sniffer-annotations:pom:1.14 from/to central (https://repo.maven.apache.org/maven2): /root/.m2/repository/org/codehaus/mojo/animal-sniffer-annotations/1.14/animal-sniffer-annotations-1.14.pom.part (No such file or directory) -> [Help 1]
```
test this please

File: backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/ElasticSearchConfiguration.java [hunk 21]
```diff
@@ -75,21 +75,21 @@ public static Credential of(String username, String password) {
         }
 
         private final String username;
-        private final String password;
+        private final char [] password;
```
Comment (chibenwa) [inline backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/ElasticSearchConfiguration.java]: `char[]`

Comment (trantienduchn) [reply backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/ElasticSearchConfiguration.java]: okay I rebased

File: backends-common/elasticsearch/src/test/resources/auth-es/README.md [hunk 22]
```diff
@@ -0,0 +1,30 @@
+## Resources explanation
+
+### nginx-conf
+
+Contains nginx configuration files:
+ - reverse_elasticsearch.conf: allow nginx to be the proxy connecting to ElasticSearch
+ - passwd: Nginx credentials file store, each record follow the format: `username:encrypted-password`
```
Comment (Arsnael) [inline backends-common/elasticsearch/src/test/resources/auth-es/README.md]: s/follow/follows

File: backends-common/elasticsearch/src/test/resources/auth-es/README.md [hunk 23]
```diff
@@ -0,0 +1,30 @@
...
+
+### default.crt & default.key
+
+public (.crt) and private (.key) of the self signed SSL certification. It will be loaded by nginx
+
+### server.jks
+
+Once you use a http client connect to nginx by `https` protocol,
```
Comment (Arsnael) [inline backends-common/elasticsearch/src/test/resources/auth-es/README.md]: s/connect/connecting

File: dockerfiles/run/guice/cassandra-ldap/destination/conf/elasticsearch.properties [hunk 24]
```diff
@@ -25,8 +25,22 @@
 elasticsearch.masterHost=elasticsearch
 elasticsearch.port=9200
 
-# Optional. Only http or https are accepted, default is http
-# elasticsearch.hostScheme=http
+# Optional, default is `default`
+# Choosing the SSL check strategy when using https scheme
+# default: Use the default SSL TrustStore of the system.
+# ignore: Ignore SSL Validation check (not recommended).
+# override: Override the SSL Context to use a custome TrustStore containing ES server's certificate.
```
Comment (Arsnael) [inline dockerfiles/run/guice/cassandra-ldap/destination/conf/elasticsearch.properties]: s/custome/custom

File: dockerfiles/run/guice/cassandra-rabbitmq-ldap/destination/conf/elasticsearch.properties [hunk 25]
```diff
@@ -27,6 +27,23 @@ elasticsearch.port=9200
 # Optional. Only http or https are accepted, default is http
 # elasticsearch.hostScheme=http
 
+# Optional, default is `default`
+# Choosing the SSL check strategy when using https scheme
+# default: Use the default SSL TrustStore of the system.
+# ignore: Ignore SSL Validation check (not recommended).
+# override: Override the SSL Context to use a custome TrustStore containing ES server's certificate.
```
Comment (Arsnael) [inline dockerfiles/run/guice/cassandra-rabbitmq-ldap/destination/conf/elasticsearch.properties]: s/custome/custom

File: dockerfiles/run/guice/cassandra-rabbitmq/destination/conf/elasticsearch.properties [hunk 26]
```diff
@@ -24,8 +24,22 @@
...
+# Optional, default is `default`
+# Choosing the SSL check strategy when using https scheme
+# default: Use the default SSL TrustStore of the system.
+# ignore: Ignore SSL Validation check (not recommended).
+# override: Override the SSL Context to use a custome TrustStore containing ES server's certificate.
```
Comment (Arsnael) [inline dockerfiles/run/guice/cassandra-rabbitmq/destination/conf/elasticsearch.properties]: s/custome/custom

File: dockerfiles/run/guice/cassandra/destination/conf/elasticsearch.properties [hunk 27]
```diff
@@ -28,6 +28,23 @@ elasticsearch.port=9200
 # Optional. Only http or https are accepted, default is http
 # elasticsearch.hostScheme=http
 
+# Optional, default is `default`
+# Choosing the SSL check strategy when using https scheme
+# default: Use the default SSL TrustStore of the system.
+# ignore: Ignore SSL Validation check (not recommended).
+# override: Override the SSL Context to use a custome TrustStore containing ES server's certificate.
```
Comment (Arsnael) [inline dockerfiles/run/guice/cassandra/destination/conf/elasticsearch.properties]: s/custome/custom

File: src/site/xdoc/server/config-elasticsearch.xml (hunk 14 above)
Comment (Arsnael) [inline src/site/xdoc/server/config-elasticsearch.xml]: s/with the ES/ES

File: src/site/xdoc/server/config-elasticsearch.xml [hunk 28]
```diff
@@ -186,6 +190,66 @@
...
+            Which means you should configure the ES RestHighLevelClient to trust your self signed certificate.
+
+            There are two ways on client side: ignoring SSL check or configure to trust the server's certificate.
+            In case you want to ignore the SSL check, simply, just don't specify below options. Otherwise, configuring the trust
+            requires some prerequisites and they are explained in below block.
+
+            A certificate normally contains two parts: a public part in .crt file, another private part in .key file.
+            To trust the server, the client need to be acknowledged that the server's certificate is in the list of
```
Comment (Arsnael) [inline src/site/xdoc/server/config-elasticsearch.xml]: s/need/needs

File: src/site/xdoc/server/config-elasticsearch.xml [hunk 29]
```diff
@@ -186,6 +190,66 @@
...
+            and <strong>trustStorePassword</strong> with the TrustStore location and the password. ES client will accept
+            the certificate of ES service.
+        </p>
+
+        <dl>
+            <dt><strong>elasticsearch.hostScheme.https.trustStorePath</strong></dt>
+            <dd>
+              Optional. Use it when https is configured in elasticsearch.hostScheme, and sslValidationStrategy is <strong>override</strong>
+              Configure Elasticsearch rest client to use this trustStore file to recognize nginx's ssl certificate.
+              Once, you chose <strong>override</strong>, you need to specify both trustStorePath and trustStorePassword.
```
Comment (Arsnael) [inline src/site/xdoc/server/config-elasticsearch.xml]: s/Once, you chose/Once you choose

File: src/site/xdoc/server/config-elasticsearch.xml [hunk 30]
```diff
@@ -186,6 +190,66 @@
...
+            </dd>
+
+            <dt><strong>elasticsearch.hostScheme.https.trustStorePassword</strong></dt>
+            <dd>
+              Optional. Use it when https is configured in elasticsearch.hostScheme, and sslValidationStrategy is <strong>override</strong>
+              Configure Elasticsearch rest client to use this trustStore file with the specified password.
+              Once, you chose <strong>override</strong>, you need to specify both trustStorePath and trustStorePassword.
```
Comment (Arsnael) [inline src/site/xdoc/server/config-elasticsearch.xml]: idem

File: dockerfiles/run/guice/cassandra-ldap/destination/conf/elasticsearch.properties (hunk 24 above)
Comment (Arsnael) [reply dockerfiles/run/guice/cassandra-ldap/destination/conf/elasticsearch.properties]: Already resolved...

File: dockerfiles/run/guice/cassandra-rabbitmq-ldap/destination/conf/elasticsearch.properties (hunk 25 above)
Comment (Arsnael) [reply dockerfiles/run/guice/cassandra-rabbitmq-ldap/destination/conf/elasticsearch.properties]: Already resolved...

File: backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/ClientProvider.java [hunk 31]
```diff
@@ -94,7 +94,7 @@ private void configureSSLOptions(HttpAsyncClientBuilder builder) {
             try {
                 builder
                     .setSSLContext(sslContext())
-                    .setSSLHostnameVerifier(NoopHostnameVerifier.INSTANCE);
+                    .setSSLHostnameVerifier(ACCEPT_ANY_HOST);
```
Comment (rouazana) [inline backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/ClientProvider.java]: why are you still using a noop hostname verifier? Why not a real one, such as DefaultHostnameVerifier ?

Comment (trantienduchn) [reply backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/ClientProvider.java]: sorry, I missed understand you yesterday

**Comment by rouazana** [comment]
```
[33a77135ec7e3a5f0c374c69717263af3ee661f9] [ERROR] Errors: 
[33a77135ec7e3a5f0c374c69717263af3ee661f9] [ERROR]   EventDeadLettersIntegrationTest.setUp:163 » Retryable Error encountered, sched...
[33a77135ec7e3a5f0c374c69717263af3ee661f9] [ERROR]   EventDeadLettersIntegrationTest.setUp:163 » Retryable Error encountered, sched...
[33a77135ec7e3a5f0c374c69717263af3ee661f9] [ERROR]   EventDeadLettersIntegrationTest.setUp:163 » Retryable Error encountered, sched...
[33a77135ec7e3a5f0c374c69717263af3ee661f9] [ERROR]   EventDeadLettersIntegrationTest.setUp:163 » Retryable Error encountered, sched...
[33a77135ec7e3a5f0c374c69717263af3ee661f9] [ERROR]   EventDeadLettersIntegrationTest.setUp:163 » Retryable Error encountered, sched...
[33a77135ec7e3a5f0c374c69717263af3ee661f9] [ERROR]   EventDeadLettersIntegrationTest.setUp:163 » Retryable Error encountered, sched...
[33a77135ec7e3a5f0c374c69717263af3ee661f9] [ERROR]   EventDeadLettersIntegrationTest.setUp:163 » Retryable Error encountered, sched...
```
test this please

File: backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/ElasticSearchConfiguration.java [hunk 32]
```diff
@@ -128,6 +129,21 @@ static SSLValidationStrategy from(String rawValue) {
             }
         }
 
+        public enum HostNameVerifier {
+            DEFAULT,
+            NOOP;
```
Comment (rouazana) [inline backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/ElasticSearchConfiguration.java]: I guess I prefer ACCEPT_ANY_HOST name as an option, it makes the intent clearer.

**Comment by chibenwa** [comment]
test this please (after CI restart)

File: backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/ElasticSearchConfiguration.java (hunk 32 above)
Comment (trantienduchn) [reply backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/ElasticSearchConfiguration.java]: > If you can remove the last commit by configuring correctly hostnames in certificates

I don't know much about the SSL things, so I didn't think about that solution. Even when it's a good suggestion, there are two concerns for me, using default hostname verifier everytime will break the test of ignoring SSL check, second, I'm not sure the docker CI behave in the same way as local. It doesn't return hostname as localhost but an IP address. That's what I observed several days ago.

**Comment by trantienduchn** [comment]
```
[dc6a17f0f1e3f32a47c460b474aa1739e3e80d80] [INFO] Apache James :: Server :: Task :: Distributed ...... FAILURE [  02:00 h]
```
test this please

File: backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/ElasticSearchConfiguration.java (hunk 32 above)
Comment (rouazana) [reply backends-common/elasticsearch/src/main/java/org/apache/james/backends/es/ElasticSearchConfiguration.java]: > > If you can remove the last commit by configuring correctly hostnames in certificates
> 
> I don't know much about the SSL things, so I didn't think about that solution. Even when it's a good suggestion, there are two concerns for me, using default hostname verifier everytime will break the test of ignoring SSL check

Yes, but it's a test issue, but I understand this point and that's why I accept this solution.

> , second, I'm not sure the docker CI behave in the same way as local. It doesn't return hostname as localhost but an IP address. That's what I observed several days ago.

An hostname verifier works also with IP when they are in the SubjectAltName field of the certificate.

**Comment by rouazana** [comment]
merged
